op-service/eth/blob.go

```diff
@@ -69,59 +71,198 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
+	} else {
+		return fmt.Errorf("Error: length_rollup_data is too large")
+	}
+
+	offset := 0
+	var buffer []byte
+
+	// encode the first 27 + 1 bytes of data into remaining bytes of first field element
+	buffer, offset = read(0, 27, data)
+	x, offset := read(offset, 1, data)
+	encodedByte := x[0] & 0b0011_1111
+	b.write([]byte{encodedByte}, 0)
+	b.write(buffer, 5)
+	pointer := 32 // manually set the pointer to the next field element
+
+	clearBuffer(buffer)
+	buffer, offset = read(offset, 31, data)
+	y, offset := read(offset, 1, data)
+
+	encodedByte = (y[0] & 0b0000_1111) | ((x[0] & 0b1100_0000) >> 2)
+	pointer = b.write([]byte{encodedByte}, pointer)
+	pointer = b.write(buffer, pointer)
+
+	clearBuffer(buffer)
+	buffer, offset = read(offset, 31, data)
+	z, offset := read(offset, 1, data)
+	encodedByte = z[0] & 0b0011_1111
+	pointer = b.write([]byte{encodedByte}, pointer)
+	pointer = b.write(buffer, pointer)
+
+	clearBuffer(buffer)
+	buffer, offset = read(offset, 31, data)
+	encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+	pointer = b.write([]byte{encodedByte}, pointer)
+	pointer = b.write(buffer, pointer)
+
+	if offset == len(data) {
+		return nil
+	}
+
+	for fieldNumber := 1; fieldNumber < 1024; fieldNumber++ {
+		clearBuffer(buffer)
+		buffer, offset = read(offset, 31, data)
+		x, offset = read(offset, 1, data)
+		encodedByte = x[0] & 0b0011_1111
+		pointer = b.write([]byte{encodedByte}, pointer)
+		pointer = b.write(buffer, pointer)
+
+		clearBuffer(buffer)
+		buffer, offset = read(offset, 31, data)
+		y, offset = read(offset, 1, data)
+		encodedByte = (y[0] & 0b0000_1111) | ((x[0] & 0b1100_0000) >> 2)
+		pointer = b.write([]byte{encodedByte}, pointer)
+		pointer = b.write(buffer, pointer)
+
+		clearBuffer(buffer)
+		buffer, offset = read(offset, 31, data)
+		z, offset = read(offset, 1, data)
+		encodedByte = z[0] & 0b0011_1111
+		pointer = b.write([]byte{encodedByte}, pointer)
+		pointer = b.write(buffer, pointer)
+
+		clearBuffer(buffer)
+		buffer, offset = read(offset, 31, data)
+		encodedByte = ((z[0] & 0b1100_0000) >> 2) | ((y[0] & 0b1111_0000) >> 4)
+		pointer = b.write([]byte{encodedByte}, pointer)
+		pointer = b.write(buffer, pointer)
+
+		if offset >= len(data) {
+			return nil
 		}
 	}
+
 	if offset < len(data) {
 		return fmt.Errorf("failed to fit all data into blob. bytes remaining: %v", len(data)-offset)
 	}
+
 	return nil
 }
 
+func read(offset int, numBytes int, data []byte) ([]byte, int) {
+	if offset >= len(data) {
+		// If the offset is at or beyond the end of data, return a new byte array of numBytes length
+		return make([]byte, numBytes), len(data)
+	}
+
+	// Calculate the actual number of bytes to read, which may be less than numBytes
+	// if the offset is near the end of data
+	actualNumBytes := numBytes
+	if offset+numBytes > len(data) {
+		actualNumBytes = len(data) - offset
+	}
+
+	// Create a new byte array and copy the data from the original slice
+	byteArray := make([]byte, actualNumBytes)
```

Comment:
Preserve the invariant that the returned by array is always exactly "numBytes".  While it works either way, I think this makes it easier to reason about.

```suggestion
	byteArray := make([]byte, numBytes)
```